File: ports/lua/src/unicode_lua/security/covert/bidi_control_balance.lua
-- Detection of Trojan-Source-class bidi-control balance hazards
-- (CVE-2021-42574 / CVE-2021-42694).  A per-type stack yields four independent
-- sub-threats: DepthExceeded, OrphanPop, UnbalancedEmbedding, UnbalancedIsolate.
-- Positions are 0-based codepoint offsets.

local calculus = require("unicode_lua.security.calculus")
local ClassificationKind = calculus.ClassificationKind

local M = {}

-- UAX #9 §3.3.2 cap on the stack-of-stacks depth.
M.UAX_DEPTH_LIMIT = 125

function M.opens_embedding(cp)
  return cp == 0x202A or cp == 0x202B or cp == 0x202D or cp == 0x202E
end

function M.is_pdf(cp)
  return cp == 0x202C
end

function M.opens_isolate(cp)
  return cp == 0x2066 or cp == 0x2067 or cp == 0x2068
end

function M.is_pdi(cp)
  return cp == 0x2069
end

function M.is_bidi_format_control(cp)
  return M.opens_embedding(cp) or M.is_pdf(cp) or M.opens_isolate(cp) or M.is_pdi(cp)
end

-- Returns { kind, sub, bidi_positions (0-based), emb_open_count, emb_pop_count,
-- iso_open_count, iso_pop_count, max_depth }.
function M.detect(input)
  local v = {
    kind = ClassificationKind.Clear,
    sub = nil,
    bidi_positions = {},
    emb_open_count = 0,
    emb_pop_count = 0,
    iso_open_count = 0,
    iso_pop_count = 0,
    max_depth = 0,
  }
  local emb_stack = 0
  local iso_stack = 0
  local orphans = {}

  for idx = 1, #input do
    local cp = input[idx]
    if M.is_bidi_format_control(cp) then
      local i = idx - 1
      v.bidi_positions[#v.bidi_positions + 1] = i
      if M.opens_embedding(cp) then
        emb_stack = emb_stack + 1
        v.emb_open_count = v.emb_open_count + 1
        if emb_stack + iso_stack > v.max_depth then
          v.max_depth = emb_stack + iso_stack
        end
      elseif M.is_pdf(cp) then
        v.emb_pop_count = v.emb_pop_count + 1
        if emb_stack > 0 then
          emb_stack = emb_stack - 1
        else
          orphans[#orphans + 1] = i
        end
      elseif M.opens_isolate(cp) then
        iso_stack = iso_stack + 1
        v.iso_open_count = v.iso_open_count + 1
        if emb_stack + iso_stack > v.max_depth then
          v.max_depth = emb_stack + iso_stack
        end
      elseif M.is_pdi(cp) then
        v.iso_pop_count = v.iso_pop_count + 1
        if iso_stack > 0 then
          iso_stack = iso_stack - 1
        else
          orphans[#orphans + 1] = i
        end
      end
    end
  end

  if #v.bidi_positions == 0 then
    return v
  end

  if v.max_depth > M.UAX_DEPTH_LIMIT then
    v.kind = ClassificationKind.Hazard
    v.sub = { tag = "DepthExceeded", max_depth = v.max_depth }
    return v
  end
  if #orphans > 0 then
    v.kind = ClassificationKind.Hazard
    v.sub = { tag = "OrphanPop", positions = orphans }
    return v
  end
  if emb_stack > 0 then
    v.kind = ClassificationKind.Hazard
    v.sub = { tag = "UnbalancedEmbedding", open_count = v.emb_open_count, pop_count = v.emb_pop_count }
    return v
  end
  if iso_stack > 0 then
    v.kind = ClassificationKind.Hazard
    v.sub = { tag = "UnbalancedIsolate", open_count = v.iso_open_count, pop_count = v.iso_pop_count }
    return v
  end
  return v
end

return M

File: ports/php/src/Security/Covert/BidiControlBalance.php
<?php

declare(strict_types=1);

namespace UnicodePhp\Security\Covert;

use UnicodePhp\Security\ClassificationKind;

final class BidiControlVerdict
{
    /** @param list<int> $bidiPositions */
    public function __construct(
        public ClassificationKind $kind,
        public ?object $sub,
        public array $bidiPositions,
        public int $embOpenCount,
        public int $embPopCount,
        public int $isoOpenCount,
        public int $isoPopCount,
        public int $maxDepth,
    ) {
    }
}

final class BidiControlBalance
{
    public const UAX_DEPTH_LIMIT = 125;

    public static function opensEmbedding(int $cp): bool
    {
        return $cp === 0x202A || $cp === 0x202B || $cp === 0x202D || $cp === 0x202E;
    }

    public static function isPdf(int $cp): bool
    {
        return $cp === 0x202C;
    }

    public static function opensIsolate(int $cp): bool
    {
        return $cp === 0x2066 || $cp === 0x2067 || $cp === 0x2068;
    }

    public static function isPdi(int $cp): bool
    {
        return $cp === 0x2069;
    }

    public static function isBidiFormatControl(int $cp): bool
    {
        return self::opensEmbedding($cp) || self::isPdf($cp) || self::opensIsolate($cp) || self::isPdi($cp);
    }

    /** @param list<int> $input */
    public static function detect(array $input): BidiControlVerdict
    {
        $v = new BidiControlVerdict(ClassificationKind::Clear, null, [], 0, 0, 0, 0, 0);
        $embStack = 0;
        $isoStack = 0;
        $orphans = [];

        foreach ($input as $i => $cp) {
            if (!self::isBidiFormatControl($cp)) {
                continue;
            }
            $v->bidiPositions[] = $i;
            if (self::opensEmbedding($cp)) {
                $embStack++;
                $v->embOpenCount++;
                $v->maxDepth = max($v->maxDepth, $embStack + $isoStack);
            } elseif (self::isPdf($cp)) {
                $v->embPopCount++;
                if ($embStack > 0) {
                    $embStack--;
                } else {
                    $orphans[] = $i;
                }
            } elseif (self::opensIsolate($cp)) {
                $isoStack++;
                $v->isoOpenCount++;
                $v->maxDepth = max($v->maxDepth, $embStack + $isoStack);
            } elseif (self::isPdi($cp)) {
                $v->isoPopCount++;
                if ($isoStack > 0) {
                    $isoStack--;
                } else {
                    $orphans[] = $i;
                }
            }
        }

        if ($v->bidiPositions === []) {
            return $v;
        }
        if ($v->maxDepth > self::UAX_DEPTH_LIMIT) {
            $v->kind = ClassificationKind::Hazard;
            $v->sub = (object) ['tag' => 'DepthExceeded', 'maxDepth' => $v->maxDepth];
        } elseif ($orphans !== []) {
            $v->kind = ClassificationKind::Hazard;
            $v->sub = (object) ['tag' => 'OrphanPop', 'positions' => $orphans];
        } elseif ($embStack > 0) {
            $v->kind = ClassificationKind::Hazard;
            $v->sub = (object) ['tag' => 'UnbalancedEmbedding', 'openCount' => $v->embOpenCount, 'popCount' => $v->embPopCount];
        } elseif ($isoStack > 0) {
            $v->kind = ClassificationKind::Hazard;
            $v->sub = (object) ['tag' => 'UnbalancedIsolate', 'openCount' => $v->isoOpenCount, 'popCount' => $v->isoPopCount];
        }
        return $v;
    }
}
